proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+
+Imports
+^^^^^^^
+
+Imports at the top level of a file module are treated specially. Their
+target is interpreted as a *dotted* name, which is resolved using the
+current rules for locating external Haskell modules. [#]_
+
+.. [#] For imports in other settings, see `Local modules`_.
+
+Importing a module brings all the exported bindings of the target module
+into scope, as described in [hsmods]_. It also brings
+the target module into scope bound to its declared simple name (which may be
+a dotted name if it is a file module).
+
+If a module is imported ``qualified`` then the target module is no
+brought into scope.
+
+If a module is imported ``as`` a simple name, then the target module is brought
+into scope with that simple name instead.
+
+If multiple modules are imported ``as`` the same name, then we retain the current
+behaviour, which is to “merge” modules. [#]_ The semantics is roughly that:
+
+.. [#] See `What should happen if multiple modules are defined
+       with the same name?`_ for discussion.
+
+::
+
+   import A as M
+   import B as M
+
+means the same as
+
+::
+
+   module M (module A(..), module B(..)) where
+       import A
+       import B
+
+This retains the current behaviour with respect to clashing names, since
+Haskell does not issue ambiguity warnings on import, only on usage. Thus
+the hypothetical definition of ``M`` is error-free even if ``A`` and
+``B`` export ``n``, but an attempt to use ``M.n`` will result in an
+ambiguity error. [#]_
+
+.. [#] This explication of the semantics uses a local module definition,
+   but we could wire the behaviour in as a special case if we decide
+   not to support local modules.
+
+Module export specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The previous section allows us to refer to modules by name, but not to
+export those names.
+
+We propose adding a language extension ``ExportModuleNames`` which adds the following
+export specifiers:
+
+- ``module M`` *changes* to exports the name ``M`` as a module binding.
+- ``module M(a, b, .. c)`` exports ``M`` and some of its names.
+- ``module M(..)`` exports ``M`` and all of its names. [#]_
+
+.. [#] See `Do we need additional syntax for re-exporting all bindings from a module?`_ for more discussion.
+
+This ``ExportModuleNames`` behaviour is the “right” behaviour as far as
+this proposal is concerned, but we gate it behind a language extension
+for backwards compatibility.
+
+Local modules
+~~~~~~~~~~~~~
+
+We propose adding a language extension ``LocalModules``. This allows
+module definitions within modules, with the same syntax as in [ghc283]_. Local
+modules are very like file modules, in particular:
+
+- They can have imports.
+
+  - The imports can only target modules which can be referred to
+    with a qualified name in the current scope. Their target module name is *never*
+    interpreted as a dotted name.
+
+- They can have nested module definitions.
+- They can have export specifiers.
+- They can *not* have dotted names.
```

Comment:
Why disallow dotted names?  File modules can have them, and the un-extended language has no problem re-exporting those.

Reply:
@deepfire I agree that we should constrain dotted names as much as possible. That is the defining characteristic of this proposal vs the others, and I think that the middle ground between them all isn't as coherent a design as the boundary.

Reply:
Yes, I think my philosophical position is that dotted names shouldn't exist, and dot should *always* mean accessing a module. The only reason they're in this proposal is to accommodate existing behaviour, and if we can come up with a way to get rid of them I'd be keen.

Reply:
> Yes, I think my philosophical position is that dotted names shouldn't exist, and dot should always mean accessing a module.

Do you have a sketch of how this would look?

E.g. [servant](https://hackage.haskell.org/package/servant) has `module Servant.API` and `module Servant.API.Alternative` (though not `module Servant`). How would module names, file layout, and module layout look in this situation, in an ideal world?

Reply:
Does the section "How could we make dotted names less special?" answer your question?

In short: defining a file module with a dotted name would give rise to a "virtual" parent module containing it. In an ideal world, such modules would be created per-package and be ambiguous if they appeared in multiple packages, and we would use `PackageImports` to disambiguate them. However, this would be very annoying in practice (`Data` would be massively ambiguous), and so probably we should just merge all such "virtual" parent modules across packages. (This inconsistency is further evidence to me that this position is not fully baked.)

Reply:
Okay, what about ByteString and ByteString.Lazy, then? In this situation you can't create a virtual module because it's already taken. Do you just merge the real top-level Bytestring module and the virtual one arising from the existence of ByteString.Lazy?

Reply:
Yes, in that situation we would have to merge them. Thanks, that's a pretty ugly case, and pushes me away from this idea. 

(That is: while I would *like* dots to always mean accessing a module, I think it may be impossible for Haskell and we'll have to have some kind of uneasy coexistence with dotted names regardless.)